docs/advanced_topics/sites/custom_site_model.rst

```diff
@@ -0,0 +1,57 @@
+=====================
+Custom site model
+=====================
+
+An alternate ``Site`` model can be used to add custom behaviour and
+additional fields.
+
+You need to complete the following steps in your project to do this:
+
+ - Create a new document model that inherits from
+   ``wagtail.core.models.AbstractSite``. This is where you would
+   add additional fields.
+ - Point ``WAGTAILCORE_SITE_MODEL`` to the new model.
+
+Here's an example:
+
+.. code-block:: python
+
+    # models.py
+    from wagtail.core.models import AbstractSite
+
+    class Site(AbstractSite):
+        # Custom field example:
+        registrar = models.CharField(
+            max_length=255,
+        )
+
+        admin_form_fields = Site.admin_form_fields + (
+            # Add all custom fields names to make them appear in the form:
+            'registrar',
+        )
+
+
+Then in your settings module:
+
+.. code-block:: python
+
+    # Ensure that you replace app_label with the app you placed your custom
+    # model in.
+    WAGTAILCORE_SITE_MODEL = 'app_label.Site'
+
+.. topic:: Migrating from the builtin site model
+
+    When changing an existing site to use a custom site model, no site
+    will be copied to the new model automatically. Copying old sites to the
+    new model would need to be done manually with a
+    :ref:`data migration <django:data-migrations>`.
+
+    Any templates that reference the builtin site model will still continue
+    to work as before.
```

Comment:
That's a good idea - migrating to a swappable model mid project is pretty much impossible and I think it's smart to call that out more

Reply:
While we're still creating a site in the initial data migration, I think we also need to highlight that it must remain possible for custom sites to be created from `host_name`, `root_page_id` and `default_site` - if there are other required, non-nullable fields, the data migration will most likely fail.

Reply:
I hadn't thought about that

Reply:
That is tricky - I think there are cases where people might want to require other fields.

Do we want to prevent those cases via a system check or catch the exception and spit out a warning in the migration? I prefer the warning but I could see getting the same functionality via allowing fields to be nullable and validating at the clean/save.

Reply:
I like the idea of catching the exception and raising a warning :) unfortunately we're in 'migration land' so any custom validation behaviour on the model won't be triggered here.

Reply:
Yeah that would be the intent - that way you could still force the field being required while still allowing the initial migration to succeed.